cni/pkg/install/install.go

```diff
@@ -51,34 +51,39 @@ func NewInstaller(cfg *config.InstallConfig, isReady *atomic.Value) *Installer {
 // Run starts the installation process, verifies the configuration, then sleeps.
 // If an invalid configuration is detected, the installation process will restart to restore a valid state.
 func (in *Installer) Run(ctx context.Context) (err error) {
-	for {
-		if err = copyBinaries(
-			in.cfg.CNIBinSourceDir, in.cfg.CNIBinTargetDirs,
-			in.cfg.UpdateCNIBinaries, in.cfg.SkipCNIBinaries); err != nil {
-			cniInstalls.With(resultLabel.Value(resultCopyBinariesFailure)).Increment()
-			return
-		}
+install:
+	if err = copyBinaries(
+		in.cfg.CNIBinSourceDir, in.cfg.CNIBinTargetDirs,
+		in.cfg.UpdateCNIBinaries, in.cfg.SkipCNIBinaries); err != nil {
+		cniInstalls.With(resultLabel.Value(resultCopyBinariesFailure)).Increment()
+		return
+	}
 
-		if in.saToken, err = readServiceAccountToken(); err != nil {
-			cniInstalls.With(resultLabel.Value(resultReadSAFailure)).Increment()
-			return
-		}
+	if in.saToken, err = readServiceAccountToken(); err != nil {
+		cniInstalls.With(resultLabel.Value(resultReadSAFailure)).Increment()
+		return
+	}
 
-		if in.kubeconfigFilepath, err = createKubeconfigFile(in.cfg, in.saToken); err != nil {
-			cniInstalls.With(resultLabel.Value(resultCreateKubeConfigFailure)).Increment()
-			return
-		}
+	if in.kubeconfigFilepath, err = createKubeconfigFile(in.cfg, in.saToken); err != nil {
+		cniInstalls.With(resultLabel.Value(resultCreateKubeConfigFailure)).Increment()
+		return
+	}
 
-		if in.cniConfigFilepath, err = createCNIConfigFile(ctx, in.cfg, in.saToken); err != nil {
-			cniInstalls.With(resultLabel.Value(resultCreateCNIConfigFailure)).Increment()
-			return
-		}
+	if in.cniConfigFilepath, err = createCNIConfigFile(ctx, in.cfg, in.saToken); err != nil {
+		cniInstalls.With(resultLabel.Value(resultCreateCNIConfigFailure)).Increment()
+		return
+	}
 
+	if in.cfg.CNIFileWatchEnabled {
 		if err = sleepCheckInstall(ctx, in.cfg, in.cniConfigFilepath, in.isReady); err != nil {
 			return
 		}
 		// Invalid config; pod set to "NotReady"
 		installLog.Info("Restarting...")
+		goto install
```

Comment:
This would make container just quit, but the install-cni container still needs to do race repairing (i.e. we want it to be long running). An easy way is to make the new env variable control the install-cni to still watch in a loop, but only install the CNI for one time.

Reply:
Thanks for the review.
After `goto install`, the container will reinstall cni files and waits for fsnotify event again, it wont quit in my opinion.
I am not sure whether we should watch the file or not if set env CNI_FILE_WATCH_ENABLED to false, maybe we should only wait for `ctx.Done`?

Reply:
I built with your patch and tried it out. If we set `CNI_FILE_WATCH_ENABLED` to false, CNI container will just exit after first install, which is not expected. I think we can rename the env var to `CNI_ENABLE_REINSTALL` and only execute installation logic once if this is set to false: https://github.com/istio/istio/blob/e2d05310fe3b3daba1d9fbfd3bf0561b265c6a14/cni/pkg/install/install.go#L55-L75